### PR TITLE
INDY-2003: Implement a command AUTH_RULE to set auth constraints

### PR DESCRIPTION
**Pull request by Toktar**

### CODE REVIEW

**Comment by Toktar** [comment]
test this please

File: indy_common/authorize/auth_constraints.py [hunk 1]
```diff
@@ -145,7 +153,8 @@ def from_dict(as_dict):
 
 class ConstraintCreator:
     @staticmethod
-    def create_constraint(as_dict):
+    def create_constraint(input_dict):
+        as_dict = dict(input_dict)
```
Comment (Toktar) [inline indy_common/authorize/auth_constraints.py]: Make deepcopy() here or before sending ```input_dict``` in ```create_constraint```

Comment (anikitinDSR) [reply indy_common/authorize/auth_constraints.py]: i think, that dict() calling is enough. It's like a copy() call

Comment (Toktar) [reply indy_common/authorize/auth_constraints.py]: But there are 3 places where we need to do dict(). We can change it to one deepcopy().

File: indy_common/types.py [hunk 2]
```diff
@@ -303,6 +307,58 @@ class ClientPoolConfigOperation(MessageValidator):
     )
 
 
+class ConstraintField(MessageValidator):
+    _base_types = None
+
+    def __init__(self, constraint_entity_cls, constraint_list_cls, **kwargs):
```
Comment (ashcherbakov) [inline indy_common/types.py]: Why do we need to explicitly pass classes here if we have just 1 variant for each argument: `ConstraintEntityField` and `ConstraintListField`?

File: indy_node/server/config_req_handler.py [hunk 3]
```diff
@@ -39,10 +35,37 @@ def doStaticValidation(self, request: Request):
...
             self._doStaticValidationPoolConfig(identifier, req_id, operation)
+        elif operation[TXN_TYPE] == AUTH_RULE:
+            self._doStaticValidationAuthRule(identifier, req_id, operation)
 
     def _doStaticValidationPoolConfig(self, identifier, reqId, operation):
         pass
 
+    def _doStaticValidationAuthRule(self, identifier, reqId, operation):
+        constraint = operation.get(CONSTRAINT)
+        ConstraintCreator.create_constraint(constraint)
```
Comment (ashcherbakov) [inline indy_node/server/config_req_handler.py]: Can it throw an exception here?

File: indy_node/server/config_req_handler.py [hunk 4]
```diff
@@ -39,10 +35,37 @@ def doStaticValidation(self, request: Request):
...
+                                       "Transaction for change authentication "
+                                       "rule for {}={} must not contain field {}".
+                                       format(AUTH_ACTION, ADD_PREFIX, OLD_VALUE))
+
+        auth_key = self.get_auth_key(operation)
+
+        if auth_key not in self.write_req_validator.auth_map and \
+                auth_key not in self.write_req_validator.anyone_can_write_map:
+            raise InvalidClientRequest(identifier, reqId,
+                                       "Key '{}' is not contained in the "
```
Comment (ashcherbakov) [inline indy_node/server/config_req_handler.py]: I would make it more clear: `Unknown authorization rule: key '' is not found in authorization map`

File: indy_node/test/auth_rule_change/test_auth_rule_transaction.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+def test_auth_rule_transaction(looper,
+                               sdk_wallet_trustee,
+                               sdk_pool_handle):
+    sdk_send_and_check_auth_rule_request(looper,
+                                         sdk_wallet_trustee,
+                                         sdk_pool_handle)
+
+
+def test_auth_rule_transaction(looper,
```
Comment (ashcherbakov) [inline indy_node/test/auth_rule_change/test_auth_rule_transaction.py]: The same test name as above

File: indy_node/test/auth_rule_change/test_auth_rule_transaction.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                                             sdk_wallet_trustee,
+                                             sdk_pool_handle,
+                                             auth_action=EDIT_PREFIX)
+    e.match("InvalidClientRequest")
+    e.match("Transaction for change authentication "
+            "rule for {}={} must contain field {}".
+            format(AUTH_ACTION, EDIT_PREFIX, OLD_VALUE))
+
+
+def test_reqnack_auth_rule_add_transaction_with_wrong_format(looper,
```
Comment (ashcherbakov) [inline indy_node/test/auth_rule_change/test_auth_rule_transaction.py]: The same test name as above

File: indy_node/test/auth_rule_change/test_auth_rule_transaction.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+                                         auth_type=NYM,
+                                         field=ROLE,
+                                         new_value='*'
+                                         )
+
+
+def test_auth_rule_transaction_with_large_constraint(looper,
+                                                     sdk_wallet_trustee,
+                                                     sdk_pool_handle):
+    constraint = _generate_constraint_list(auth_constraints=[_generate_constraint_entity(role=TRUSTEE),
```
Comment (ashcherbakov) [inline indy_node/test/auth_rule_change/test_auth_rule_transaction.py]: Can we add test with even more complex constraint?

File: indy_common/types.py [hunk 8]
```diff
@@ -303,6 +307,58 @@ class ClientPoolConfigOperation(MessageValidator):
...
+        (NEED_TO_BE_OWNER, BooleanField(optional=True)),
+        (METADATA, AnyMapField(optional=True))
+    )
+
+
+class ConstraintListField(MessageValidator):
+    schema = None
+
+    def __init__(self, schema_is_strict=SCHEMA_IS_STRICT):
+        self.schema = (
```
[the remark this answers is not in the document]
Comment (Toktar) [reply indy_common/types.py]: As I understand it's only one way to use a self class name in a schema.

File: indy_common/types.py [hunk 9]
```diff
@@ -303,6 +307,58 @@ class ClientPoolConfigOperation(MessageValidator):
     )
 
 
+class ConstraintField(MessageValidator):
```
Comment (ashcherbakov) [inline indy_common/types.py]: Please add tests for `ConstraintField`, `ConstraintEntityField` and `ConstraintListField`

**Comment by Toktar** [comment]
test this please

File: indy_common/types.py [hunk 10]
```diff
@@ -334,22 +336,20 @@ class ConstraintEntityField(MessageValidator):
...
+              (AUTH_CONSTRAINTS, IterableField(AnyField())))
 
-    def __init__(self, schema_is_strict=SCHEMA_IS_STRICT):
-        self.schema = (
-            (CONSTRAINT_ID, ChooseField(values=ConstraintsEnum.values())),
-            (AUTH_CONSTRAINTS, IterableField(ConstraintField(self,
-                                                             ConstraintEntityField())))
-        )
-        super().__init__(schema_is_strict)
+    def _validate_message(self, val):
```
Comment (ashcherbakov) [inline indy_common/types.py]: It looks like we don't do any specific validation for `AUTH_CONSTRAINTS`, so it means that list may contain any values and validation will pass?

File: indy_node/test/auth_rule_change/test_constraint_field.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    invalid_auth_rule_operation = generate_auth_rule_operation(constraint=generate_constraint_list(
+        auth_constraints=[generate_constraint_entity(),
+                          generate_constraint_list(
+                              auth_constraints=[generate_constraint_entity(),
+                                                generate_constraint_entity()])]))
+    del invalid_auth_rule_operation[CONSTRAINT][AUTH_CONSTRAINTS][1][AUTH_CONSTRAINTS]
+    with pytest.raises(TypeError) as e:
+        validator.validate(invalid_auth_rule_operation)
+    e.match(AUTH_CONSTRAINTS)
+
```
Comment (ashcherbakov) [inline indy_node/test/auth_rule_change/test_constraint_field.py]: Please add a test for presence of `CONSTRAINT_ID ` field

File: indy_node/test/auth_rule_change/test_constraint_field.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    invalid_auth_rule_operation = generate_auth_rule_operation(auth_action="auth_action")
+    with pytest.raises(TypeError) as e:
+        validator.validate(invalid_auth_rule_operation)
+    e.match(AUTH_ACTION)
+
+
+def test_invalid_entity_role():
+    # ConstraintEntityField without required field 'role'
+    invalid_auth_rule_operation = generate_auth_rule_operation()
+    del invalid_auth_rule_operation[CONSTRAINT][ROLE]
```
Comment (ashcherbakov) [inline indy_node/test/auth_rule_change/test_constraint_field.py]: Please add a test for validation of constraint in extra_large case (remove role from a second-level consraint)

**Pull request merged**